File: Loesungen/library.cpp
#include "library.h"
#include <iostream>
#include <math.h>
#include <vector>
#include <bitset>
#include <functional>
#include <random>
#include <algorithm>
#include <array>
#include <numeric>
#include <string_view>
#include <cstdint>

int gcd(int a, int b)
{
    std::cout << "a: " << a << " b: " << b << std::endl;
    if (b == 0) {
        //std::cout << "\tReturn \'" << a << "\'" <<std::endl;
        return a;
    }
    else {
        //std::cout << "\tReturn \'" << a % b << "\'" << std::endl;
        return gcd(b, a % b);
    }
}

int  lcm(int a, int b)
{
    int h = gcd(a, b);
    return h ? (a * (b / h)) : 0;
}


bool isPrim(int limite)
{
    if (limite<=3)
        return (limite>1);
    else if ((limite%2==0)||(limite%3==0)){
        return false;
    }
    else
    {
        int upLim=floor(sqrt(limite))+1;
        for (int i=2;i<upLim;i++)
        {
            if (limite%i==0)
            {
                return false;
            }
        }
    }
    return true;
}

int sum_proper_divisors(int num){
    int result=1;
    for (int i=2;i<std::sqrt(num);i++){
        if (num%i==0)
        {
            result+=(i==(num%i)?i:(i+num/i));
        }
    }
    //std::cout << num << " Sum " << result<< std::endl;
    return result;
}

void print_abundant(int limit){
    for (int number=10;number<=limit;number++)
    {
        auto sum = sum_proper_divisors(number);
        if (sum>number){
            std::cout <<number << ", Abundance " << sum-number << std::endl;
        }
    }
    return;
}

void print_amicables(int limit){
    for (int number=4;number<=limit;number++)
    {
        auto sum1 = sum_proper_divisors(number);
        if (sum1 < limit)
        {
            auto sum2 = sum_proper_divisors(sum1);
            if ((sum2==number) && (number!=sum1))
            {
                std::cout << number << "," << sum1 << std::endl;
            }
        }
    }
    return;
}

void print_narcissistics(void)
{
    for (int i=1;i<=9;i++)
    {
        for (int j=1;j<=9;j++)
        {
            for (int k=1;k<=9;k++)
            {
                auto abc =100*i+10*j+k;
                auto arm = i*i*i+j*j*j+k*k*k;
                if (abc==arm)
                {
                    std::cout << arm << std::endl;
                }
            }
        }
        //std::cout << i << std::endl;
    }
}

std::vector<unsigned long long> prime_factors(unsigned long long number)
{
    std::vector<unsigned long long> factors;
    while (number%2==0) {
        factors.push_back(2);
        number /= 2;
    }

    for (unsigned long long i=3;i<=std::sqrt(number); i++)
    {
        while (number%i==0){
            factors.push_back(i);
            number /= i;
        }
    }
    if (number>2){
        factors.push_back(number);
    }

    return factors;
}

void print_vector(std::vector<unsigned long long int> vector)
{
    std::cout << "[ ";
    for (auto i:vector){
        std::cout << i << ", ";
    }
    std::cout << "]" << std::endl;
}

unsigned int gray_encode(unsigned int const num)
{
    return num ^ (num>>1);
}

unsigned int gray_decode(unsigned int gray)
{
    for (unsigned int bit=1<<31;bit>1;bit>>=1)
    {
        if (gray & bit) gray ^= bit>>1;
    }
    return gray;
}

std::string to_binary(unsigned int value, unsigned int digits)
{
    return std::bitset<32>(value).to_string().substr(32- digits, digits);
}

std::string to_roman(unsigned int number)
{
    std::vector<std::pair<unsigned int, char const*>> roman
    {
        { 1000, "M" },{ 900, "CM" },
        { 500, "D" },{ 400, "CD" },
        { 100, "C" },{ 90, "XC" },
        { 50, "L" },{ 40, "XL" },
        { 10, "X" },{ 9, "IX" },
        { 5, "V" },{ 4, "IV" },
        { 1, "I" }
    };
    printf("Enter\n");
   std::string result;
   for (auto const & kvp : roman)
   {
      while (number >= kvp.first)
      {
         result += kvp.second;
         number -= kvp.first;
      }
   }

   return result;
}

std::pair<unsigned long long, long> longest_collatz(unsigned long long limit)
{
    long length = 0;
    unsigned long long number =0;

    std::vector<int> cache(limit+1,0);

    for (unsigned long long i=2;i<= limit; i++)
    {
        auto n = i;
        long steps = 0;
        while (n != 1 && n >= i)
        {
            if ((n % 2) == 0) n = n / 2;
            else n = n * 3 + 1;
            steps++;
        }
        cache[i] = steps + cache[n];

        if (cache[i] > length)
        {
            length = cache[i];
            number = i;
        }
    }
    return std::make_pair(number, length);
}


bool validate_isbn_10(std::string isbn)
{
   auto valid = false;
   if (isbn.length() == 10)
   {
       int w=10;
       int sum = 0;
       for (auto d = isbn.cbegin(); d != isbn.cend();d++)
       {           
           //std::cout << "Check " <<  *d << std::endl;
           if (false == isdigit(*d)) {
               printf("No digit %c",*d);
               return false;
           }
           sum += w*(*d-'0');
           w--;
       }
       valid = (sum%11)==0;
    } else {
        printf("Length not correct\n");
    }

   return valid;
}

unsigned int factorial(unsigned int n) {
    unsigned int res = 1;
    
    for (unsigned int i=1;i<=n;i++)
    {
        res *= i;
    }
    return res;
    
}

unsigned int binomial_koeffizent(unsigned int  n, unsigned int k){
    unsigned int res =0;
    if (k<=n)
    {
        unsigned int zaehler = factorial(n);
        unsigned int nenner =factorial(k)*factorial(n-k);
        res = zaehler/nenner;
    }
    //std::cout << "n = " << n << " k = " << k << " -> " << res << std::endl;
    return res;
}
    
    

unsigned int num_of_digits(unsigned int nr){
    if (nr>0){
        int32_t dCnt = ceil(log10(double(nr)));
        //std::cout << "bin= " <<nr << " digitcout  " << dCnt << std::endl;
        return dCnt;
    }else {
        return 1;
    }
}
   
std::string einruecken(uint32_t cnt){
    std::string line;
   for (uint32_t i=0;i<cnt;i++)
    {
        line.append(" ");
    }
    return line;
}

std::string fixed_width_i32(int32_t val, uint8_t width)
{
    std::string res = " ";
    if (val >0) {
        res = std::to_string(val);
    }
    while (res.length() <= width)
    {
        if (res.length()%2==0)
        {
            res += " ";
        } else {
            res = " "+ res;
        }
    }
    //std::cout << "??" << res.length() << "??" << std::endl;
    return res+ " ";
}

std::string pascal_dreieck(int32_t count_of_lines)
{
    std::string lines="";
    unsigned int width = 2+ num_of_digits(binomial_koeffizent(count_of_lines,count_of_lines>>1));
    //std::cout << "max bin= " << binomial_koeffizent(count_of_lines,count_of_lines>>1) << std::endl;
    width = 4;
    for (int n=0;n<count_of_lines; n++)
    {
        int32_t indent = (count_of_lines * width) - width * n + width;
        std::string line = einruecken(indent);
        std::cout << indent << " " << width << std::endl;
        for (int32_t k=0;k<=n;k++)
        {
            //std::cout << "    " << indent << " " << std::endl;
            uint32_t binCoef = int(binomial_koeffizent(n, k));
            line += fixed_width_i32(binCoef, width);
            //line += fixed_width_i32(-1, width);

        }
        line+= "\n";
        lines.append(line);
    }
    
    return lines;        
    
}

File: Loesungen/library.py

import numpy as np
from math import *
import random
import optparse
import struct
import os


def gcd(a,b):
    if b==0:
        return a
    else:
        return gcd(b, a % b)


def lcm(a,b):
    h = gcd(a,b)
    if h:
        return (a*(b/h))
    else:
        return 0

def isPrim(lim):
    if (lim<3):
        return lim >1
    else:
        checkLim=int(ceil(sqrt((lim))))
        #print("Check up to %d" % (checkLim))
        for i in range(2,checkLim+1):
            #print("Check %d" %i)
            if lim%i==0:
                return False
    return True


def sum_proper_divisors(num):
    res = 1
    for i in range(2,int(ceil(sqrt(num)))):
        #print("Check %d" %i)
        if num%i==0:
            res+=i if (i==num/i) else (i+num/i)
    #print("num: %d Sum %d" %(num, res))
    return res


def print_amicables(limit):
    for  number in range(4, limit):
        sum1 = sum_proper_divisors(number);
        if (sum1 < limit):
            sum2 = sum_proper_divisors(sum1);
            if ((sum2==number) and (number!=sum1)):
                print("%d, %d" % (number, sum1))


def print_armstong():
    for i in range(1,10):
        for j in range(1,10):
            for k in range(1,10):
                abc=100*i+10*j+k
                abm=i*i*i+j*j*j+k*k*k
                if  abc==abm:
                    print("%d ; %d ; %d ; %d" %(abm, i, j, k))
def getOpt():
    p = optparse.OptionParser()
    opt, args = p.parse_args()
    return opt, args

def prime_factors(number):
    factors=[]
    while (number%2==0):
        factors.append(2)
        number/=2
    uplim=int(ceil(sqrt(number)))+1
    for i in range(3,uplim,2):
        while (number%i==0):
            factors.append(i)
            number=number/i
    if number>2:
        factors.append(number)
    return factors
        

def gray_encode(number):
    return number^(number>>1)


def gray_decode(gray):
    for  bitPos in range(31,1,-1):
        bitMask = 1<<bitPos
        if gray&bitMask:
            gray ^= bitMask >>1
    return gray

def to_binary(number, digits=5):
    mask= (1<<digits)-1
    number &= mask
    string=""
    for i in range(digits):
        if number&(1<<i)==0:
            string = "0"+string
        else:
            string = "1"+string
    return string

def to_roman(number):
    if number>=10000:
        return "Not possible"
    else:
        table=(
            (1000, "M"),
            (900 , "CM"),
            (500, "D"),
            (400, "CD"),
            (100, "C"),
            (90, "XC"),
            (50, "L"),
            (10, "X"),
            (9, "IX"),
            (5, "V"),
            (4, "IV"),
            (1 , "I")            
        )
        res=""
        for t in table:
            while number >= t[0]:
                number -= t[0]
                res += t[1]
        return res
        
def longest_collatz(limite):
    cache = [0]*(limite+1)
    length = 0
    for i in range(2, limite+1):
        n = i
        steps = 0
        while ((n!=1) and (n>=i)):
            if (n%2 == 0):
                n = int(n/2)
            else:
                n = int(3*n+1)
            steps+=1
        cache[i] = steps + cache[n]
        if cache[i]>length:
            length =  cache[i]
            number = i
   
    return number, length
    
def compute_pi(samples = 1000000):
    hit = 0
    for i in range(samples):
        x = random.random()
        y = random.random()
        if y<sqrt(1-x**2):
            hit += 1
    return 4.0*hit/samples

def validate_isbn_10(isbn):
    #print(type(isbn), isbn)
    if not type(isbn)==str:
        print("Wrong datatype")
        return False
    if len(isbn) != 10:
        print("Wrong size")
        return False
    if not isbn.isdigit():
        print("Is not number")
        return False
    sum = 0
    for (w,d) in zip(range(10,0,-1), isbn):
        sum += w*(ord(d)-ord('0'))
        #print(w,ord(d)-ord('0'), sum%11)
    return (sum%11==0)

class ip4iter:
    def __init__(self, minimum, maximum):
        print(type(minimum), type(maximum))
        self._idx = int(minimum)
        self._end = int(maximum)
        print("Set miniumum to %d" % self._idx)
        print("Set maximum¨ to %d" % self._end)

    def __iter__(self):
        return self

    def __next__(self):
        self._idx += 1
        if self._idx < self._end:
            return ipv4(self._idx)
        raise StopIteration

class ipv4:
    def __init__(self, ip = None):
        self._ip = [0,0,0,0]
        self._ipu32 = 0
        if ip is None:
            return None 
        if type(ip) == type([] or type(ip) == type(())) and length(ip)==4:
            print("Is tuple ", ip)
            for i in range(4):
                self._ip[i]= ip[i] 
        elif type(ip) == int:
            self._ipu32 = ip
            self._ip = [(ip>>24)&0xFF, (ip>>16)&0xFF, (ip>>8)&0xFF, ip&0xFF]
        self._ipu32 = int.from_bytes(struct.pack("BBBB", *self._ip), byteorder='big', signed=False)

    def __eq__(self, other_ip):
        if type(other_ip) == type(self):
            self._ip= ipv4(other_ip)

    def __iter__(self):
        return self.__

    def __str__(self):
        res = "[%d.%d.%d.%d] (%d)" %(self._ip[0], self._ip[1], self._ip[2], self._ip[3], self._ipu32 )
        return res

    def readIp(self):
        ip = input("IP in dot notation ")
        ip_split = ip.split(".")
        if len(ip_split) != 4:
            print("Dot notation contains 4 values")
        self._ip = [int(i) for i in ip_split]
        self._ipu32 = (self._ip[0]<<24) + (self._ip[1]<<16) + (self._ip[2]<<8) + self._ip[3]
        print(self)

    def __add__(self, nr):
        self._ipu32 + nr

    def __int__(self):
        return self._ipu32
    
class array2d:
    def __init__(self, xSize, ySize, dtype=np.int32):
        self._arr= np.zeros((xSize, ySize),dtype=dtype)
        
    def set(self, other):
        if type(other) == type(self):
            for i in range(self._arr.shape[0]):
                for j in range(self._arr.shape[1]):
                    self._arr[i][j] = other.arr[i][j]
        if type(other) != type([]):
            raise ValueError("Wrong datatype given, must be list")
        if len(other) != self._arr.shape[0]*self._arr.shape[1]:
            errStr="Wrong vector size, given, must be %d" % (self._arr.shape[0]*self._arr.shape[1])
            raise ValueError(errStr)
        idx =0
        for i in range(self._arr.shape[0]):
            for j in range(self._arr.shape[1]):
                self._arr[i][j] = other[idx]
                idx+=1
    
    def __str__(self):
        res="[\n"
        for i in range(self._arr.shape[0]):
            for j in range(self._arr.shape[1]):
                if j==0:
                    res += " ["
                res += "%d " %self._arr[i][j]
            res += "]\n"
        res += "]\n"
        return res
    
    @property
    def size(self):
        return self._arr.shape
    
    @property
    def arr(self):
        return self._arr
    
    def fill(self,value):
        for i in range(self._arr.shape[0]):
            for j in range(self._arr.shape[1]):
                self._arr[i][j] = value
                
    def swap(self, other):    
        if type(other) != type(self):
            raise ValueError("Wrong datatype, other must be %s" % (type(self)))
        for i in range(self._arr.shape[0]):
            for j in range(self._arr.shape[1]):
                a = self._arr[i][j]
                b = other.arr[i][j]
                other.arr[i][j] = a
                self._arr[i][j] = b 
            
    def move(self, other):
        if type(other) != type(self):
            raise ValueError("Wrong datatype, other must be %s" % (type(self)))
        for i in range(self._arr.shape[0]):
            for j in range(self._arr.shape[1]):
                self._arr[i][j] = other.arr[i][j]

def factorial(n):
    res = 1
    for i in range(2,n+1):
        res *= i
    return res
    


def binomial_koeffizent(n, k):
    #print("%d tief %d " %(n,k), end= "")
    if k>n:
        res = 0
    else:
        zaehler = factorial(n)
        nenner =factorial(k)*factorial(n-k)
        res = zaehler/nenner
    #print(res)
    return res
    
    

def num_of_digits(nr):
    if nr>0:
        return int(log10(nr)+1)
    else:
        return 1
   
def pascal_dreieck(line_cnt, scan=True):
    n = num_of_digits(binomial_koeffizent(line_cnt,line_cnt>>1))
    lines=[((line_cnt<<1)*" ")+"1"]
    fmt = "{:<%d} "%n
    for i in range(1, line_cnt):
        line = ' '*((line_cnt<<1)-(2*i))
        for n in range(0, i+1):
            res = int(binomial_koeffizent(i, n))
            line += fmt.format(res )
        line+= "\n"
        lines.append(line)
    return lines        
        
def recursive_size(root_folder, res=None):
    if res ==None:
        res = {root_folder:0}
    
    for root, subdirs, files in os.walk(root_folder, True): 
        if root not in res:
            res[root]=0
        file_size = 0
        for f in files:
            fname = os.path.join(root, f)
            file_stat = os.stat(fname)
            file_size += file_stat.st_size
        res[root]= file_size
        print("Folder %s size is %d" % (root, file_size))
        for d in subdirs:
            subFolderName = os.path.join(root, d)
            print("SubDir %s"% (subFolderName))
            res = recursive_size(subFolderName, res)
        return res
    